### ISSUE TITLE
Metaserver requires a lot of CPU sometimes

### ISSUE THREAD

**Issue by SirVer** [widelands/widelands_metaserver#38]
I just was logged into the server and saw that it reported a load of > 3. This seemed plenty for the workload it should be doing so I decided to look into top:

The wlms process ate between 20-30% cpu consistently. I logged into the metaserver and saw that a game was running, so that might be an explanation (but still seems pretty high). The `wlnr` process did also run, but did not need any CPU at all, which seemed peculiar and let's me wonder if the wlms is always as resource hungry.

The server has a single CPU and mysqld is always eating on that performance too, so getting the metaserver to use fewer resources would be great.

**Comment by SirVer** [comment]
verifying: once the game ended, CPU usage was still very high. I restarted the process and the CPU usage dropped very strongly.

**Comment by SirVer** [comment]
One potentially important: there was a weird account `Rosa<long string of numbers>` that often disconnected and reconnected with a different set of numbers at the end. I wonder if that was a bot & an attacker was trying something against the metaserver. Of course this could just be random circumstance.

**Comment by SirVer** [comment]
Another update: I just realized that after restarting the process, `mysqld`'s cpu usage also dropped very strongly. Since the wlms only talks to mysqld on login, I feel this is giving more credibility to a concerted attack: somebody trying to brute force passwords potentially.

**Comment by SirVer** [comment]
`sudo zgrep Rosa wlmetaserver.log wlmetaserver.log.*.gz` lists the name 4961 times, the first time in September 12 2018. SInce then, the name appears daily in our logs.

- The first occurences of that id are `Rosa`, then `Rosa1`. 
- The numbers increase, but monotonically, but with large gaps.
- On Sept 17, the pattern shifts from `Rosa2089809280` to `Rosa-2144026275`. This seems like a i32 overflow to me.

My interpretation: Someone is either on purpose or accidentally running a bot that very aggressively connects to our service. A workaround we should implement immediately is sleeping for 1s after any login attempt: users that login will not care significantly, but it should stop ddos cold. We could also implement IP blacklisting.

**Comment by SirVer** [comment]
Interesting though: in the logs I only find login information for Client Rosa and Client Rosa1 logged in, so the other numbers never logged in. Still they are displayed in the pending user list.

So another explanation could be this: we do not clean up for a disconnected user for all circumstances and we try to login a ghost, constantly incrementing the number to distinguish them from an already existing client.

This concludes my analysis for now. I think the bug is more likely now. @Notabilis what do you think?

**Comment by Notabilis** [comment]
Thanks for the investigation. Good catch that the Rosa's never logged in, I didn't noticed that. Surely something strange is going on.
- I think your statement that Rosa#### is a ghost could be. Not that I have any idea where it is coming from, and there are some strange effects connected to it as well.
- It seems that the metaserver was trapped in the method Client::findUnconnectedName(). There it increases the number behind the playername and repeatedly checks on the SQL server whether the name is available. So it would explain the load for both processes.
- What I don't understand is why the loop never stopped. It should stop as soon as the user database and the server do not know about the tested name. In that case that would mean that all possible Rosa#### names are registered, either in the database or more likely on the metaserver. No idea why this should be the case, though.
  - When the client is already in the list of connected clients at the metaserver it would explain the loop since it is found when looking for (seemingly another) client with the same name.
  - No idea why it should be in that list, though.
- When doing ... something (logging in, starting game) on the metaserver, Rosa logged in and another Rosa logged out. I have no idea where this messages are coming from if the code is trapped inside the mentioned loop. And even apart from that it does not explain why it "reacted" to events on the server. (Due to this behavior I expected Rosa to be a poorly written bot that crashed and restarted on unexpected network messages.)
- The metaserver tried to send messages to Rosa, which means its state was CONNECTED. However, this couldn't been (?) when still trying to find a free name.
- I think I found a code path where logging out / removing the user from the metaserver is indeed be missing. It doesn't explain the occurrences (yet), though. Also, the logged in Rosa and Rosa1 show correct disconnects later on so they should have been removed. Probably this is an unrelated bug(?).

Things to do:
- Try to figure out what was going on. Right now there are multiple observations/interpretations which just doesn't fit together (yet).
- Add a maximum value for the number of names that are tried, e.g., only increase the counter to 1000 and then abort. That should never be reached, but seemingly the method looped in this case.

Addendum:
Just noticed the logs when the first strange Rosa appeared:
  - Client Rosa logged in
  - Error while handling command GAME_CONNECT for client Rosa: NO_SUCH_GAME
  - Client Rosa logged in
  - Client Rosa left. Reason: 'CONNECTION_CLOSED'
  - Warning: Error while sending data to client Rosa88656
  - Removing client Rosa

Rosa was logged in and for some reason did / was able to do a second login. No idea how this is possible, but it seems like it leads to problems.

**Comment by SirVer** [comment]
> What I don't understand is why the loop never stopped. It should stop as soon as the user database and the server do not know about the tested name. 

I concur that this is peculiar. Skimming the code it seems that both are very unlikely to be wrong, i.e. the implementation of the SqlDatabase and the HasClient implementation seem to be correct and surrounding code free of race conditions. I am not completely sure though - are we running the server in a single thread? If not, there might be race conditions in modifying the lists.

I suggest adding more logging into the function to figure out which code path is the problem should the issue arise again. It might also be useful to log more of the state machine a user goes through, i.e. logging in, sending messages, leaving so that the logs show a more complete picture.

About your Addendum: this smells like a race condition to me. I still cannot explain how it leads to the infinite loop that we saw, but race conditions could lead to invalidated memory potentiallly, so everything can happen. We could look into instrumentation to catch this, but I have no idea how much work that is: https://medium.com/@val_deleplace/does-the-race-detector-catch-all-data-races-1afed51d57fb

**Comment by Notabilis** [comment]
The code uses a lot of goroutines, so I guess we could have some race conditions. Good pointer actually, I should look after them while trying to debug the code.

My current hypothesis about the loop is that the client is already in the list of connected clients while searching for a free name. Since it overwrites its name to compare it with other clients, it finds itself while iterating through the list of connected clients. Since it has the same name as itself it considers the name taken and tries the next one. This would explain the loop, but I still have to figure out why it is already in the list, possibly due to the NO_SUCH_GAME error.

I looked in the client code and the send error about NO_SUCH_GAME is ignored on the client. Instead, waiting for the result of GAME_CONNECT should timeout after a few seconds and the client reconnects. I think that might be what happens in the above log extract. I haven't followed through the metaserver code when this happens yet (will be my next step). What is a bit strange is that the client should send another GAME_CONNECT after reconnect, but that doesn't seem to happen here.

**Comment by Notabilis** [comment]
Unfortunately I have only partially been successful debugging this. I have an idea what has happened but I can't explain how that state has been reached. I managed to create a metaserver log equivalent to the one above, but I had to modify my Widelands client for it and I doubt that this happened in the above case. My assumption about what happened:

- Rosa did a normal login and tried to join a game
- For some reason (probably network delay, so no bug here) the game was no longer available when the "JOIN" command reached the metaserver
- The metaserver responded with an error message that the game is no longer available
- Now the behavior differs between the expected behavior and the above log
  - Normally, the client should ignore the error message, timeout and then reconnect and try again a few times before giving up
  - Instead, I modified my client to *not* reconnect but send another LOGIN request over the same TCP connection. Since the client is already in the list of clients, it leads to the observed loop.

Due to my modifications I had one TCP connection between client and metaserver but two metaserver "threads" handling it: A normal one that handled commands correctly and a second goroutine looping while trying to find an empty username but always encountering itself. That fits what have been observed on the server but I wasn't able to figure out how it could have happened with an unmodified client (based on previous behavior of Rosa, I don't think it has been a deliberate attack), even taking strange timing issues into consideration.

Anyway, I pushed a branch that should avoid this effect in the future, even though I am unfortunately unable to explain how it happened in the first place.
